File: lclang/src/main/java/lclang/statements/expressions/GetTypeExpression.java
package lclang.statements.expressions;

import lclang.Caller;
import lclang.LCBaseExecutor;
import lclang.Link;
import lclang.exceptions.LCLangRuntimeException;
import lclang.libs.lang.classes.StringClass;

public class GetTypeExpression extends Expression {
    public final Expression expression;

    public GetTypeExpression(Expression expression) {
        super(0);
        this.expression = expression;
    }

    @Override
    public Link visit(Caller prevCaller, LCBaseExecutor visitor) throws LCLangRuntimeException {
        return StringClass.get(expression.visit(prevCaller, visitor).get
                .invoke(prevCaller).classType.text).createLink();
    }
}

File: lclang/src/main/kotlin/lclang/statements/GetTypeExpression.kt
package lclang.statements

import lclang.Caller
import lclang.LCBaseExecutor
import lclang.libs.lang.StringClass

class GetTypeExpression(
    val of: Expression,
    line: Int,
    column: Int
): Expression(line, column) {
    override fun visit(prevCaller: Caller, visitor: LCBaseExecutor) =
        StringClass(of.visit(getCaller(prevCaller), visitor).type.text).asValue()
}